codebase/src/java/disco/org/openlvc/disco/utils/Quaternion.java

```diff
@@ -72,19 +76,39 @@ public Quaternion multiply( Quaternion q )
 	 * 
 	 * @return an {@link EulerAngles} instance suitable for use in a DIS PDU for setting entity
 	 *         orientation
+	 * 
+	 * @see #fromPduEulerAngles(EulerAngles)
 	 */
 	public EulerAngles toPduEulerAngles()
 	{
-		double sqw = w * w;
-		double sqx = x * x;
-		double sqy = y * y;
-		double sqz = z * z;
-		
-		double yaw = Math.atan2( 2.0 * (x * y + z * w), (sqx - sqy - sqz + sqw) );
-		double roll = Math.atan2( 2.0 * (y * z + x * w), (-sqx - sqy + sqz + sqw) );
-		double pitch = Math.asin( -2.0 * (x * z - y * w) / (sqx + sqy + sqz + sqw) );
+		double pitchRatio = -2.0 * (x * z - y * w);
+
+		double divA = w * w - y * y;
+		double divB = x * x - z * z;
+
+		double yaw = Math.atan2( 2.0 * (x * y + z * w), divA + divB );
+		double pitch;
+		double roll = Math.atan2( 2.0 * (y * z + x * w), divA - divB );
+
+		// handle pitch singularities
+		if( Math.abs( pitchRatio ) > 0.998 ) // > ~86.4 degrees
+		{
+			// yaw and roll are aligned - use only yaw
+			yaw = yaw - roll;
+			pitch = pitchRatio > 0d ? EulerAngles.THETA_MAX : EulerAngles.THETA_MIN;
```

Comment:
This is purely up to taste, but I think formatting tertiary operators the following way makes them much more readable.

```
pitch = pitchRatio > 0.0 ? EulerAngles.THETA_MAX
                         : EulerAngles.THETA_MIN;
```